### PR TITLE
Adds --use-rra for abstracting goto programs using Replication Reducing Abstraction

### PR DESCRIPTION
**Pull request by muralitaws**
This can be used to unbounded proofs using CBMC. User has to specify the data structures to be abstracted
in a json file along with use-rra flag.

Current commit introduces infrastructure for parsing the spec files, the option with a minimal flow,
and a set of regression tests.

Joint work with Lefan Zhang (lefanz@amazon.com).

- [x ] Each commit message has a non-empty body, explaining why the change was made.
- [ x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ x] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ x] My commit message includes data points confirming performance improvements (if claimed).
- [ x] My PR is restricted to a single feature or bugfix.
- [ x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Comment by muralitaws** [comment]
This PR adds the new flag with parsing infrastructure, and regression tests but doesn't really abstract the given goto-program apart from appending to it a few functions that will be used by the abstraction. A detailed documentation of the feature is appropriate in the next PR that will contain the actual functionality. I am breaking the code up into smaller PRs to keep things manageable. The documentation is currently limited to the comments in abstraction.h header file.

**Comment by martin-cs** [comment]
Do I understand correctly...

1. This PR includes test cases which are marked as core but aren't actually run by `cd regression; make test`?
2. They may or may not work because this PR doesn't include the actual main part of the functionality?
3. There is minimal documentation so for those of us who don't know what RRA is, we can't check that the test results are correct?

It looks like an exciting new feature and splitting things up is appreciated but maybe there could be an easier split to review?

File: regression/abstraction/Makefile [hunk 1]
```diff
@@ -0,0 +1,38 @@
+default: tests.log
```
Comment (martin-cs) [inline regression/abstraction/Makefile]: Please can this be `regression/replication-reducing-abstraction/Makefile` or similar? Just `abstraction` is too ambiguous.

File: regression/abstraction/abst-funcs.c [hunk 2]
```diff
@@ -0,0 +1,714 @@
+#include <assert.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+// Some helper functions.
+
```
Comment (martin-cs) [inline regression/abstraction/abst-funcs.c]: Why not just have `size_t nondet_size_t{)` in the header and use that CPROVER doesn't issue 'missing function' warnings for functions whose names start with `nondet`.

File: regression/abstraction/abst-funcs.c [hunk 3]
```diff
@@ -0,0 +1,714 @@
...
+// nndt_bool, nndt_int, nndt_sizet are available in CBMC.
+
+size_t nndt_int()
```
Comment (martin-cs) [inline regression/abstraction/abst-funcs.c]: `size_t` and `int` may be quite different.

File: regression/abstraction/abst-funcs.c [hunk 4]
```diff
@@ -0,0 +1,714 @@
...
+{
+  size_t i;
+  return (i);
+}
+
+bool nndt_bool()
+{
+  size_t i;
```
Comment (martin-cs) [inline regression/abstraction/abst-funcs.c]: Why not `bool`?

File: regression/abstraction/abstraction.sh [hunk 5]
```diff
@@ -0,0 +1,58 @@
+#!/bin/bash
```
Comment (martin-cs) [inline regression/abstraction/abstraction.sh]: If the directory is renamed `abstraction.sh` would be less of a problem but, again, it would be better with a less ambigious name.

File: regression/abstraction/abstraction.sh [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+# whether verify the program or not
+CHECK=true
+
+# executables
+bindir="$(pwd)/../../build/bin"
+GOTO_CC="$bindir/goto-cc"
+GOTO_INSTRUMENT="$bindir/goto-instrument"
+CBMC="$bindir/cbmc"
+
+CBMC_FLAGS="--unwind 4
```
Comment (martin-cs) [inline regression/abstraction/abstraction.sh]: Why 4?

File: regression/abstraction/test2/Readme [hunk 7]
```diff
@@ -0,0 +1,4 @@
+Assertions with function "bar" that has hidden index variable.
+We error out currently because functions are not supported inside assertions.
```
Comment (martin-cs) [inline regression/abstraction/test2/Readme]: If you error out, should this be a `CORE` test? `KNOWNBUG` or `FUTURE` would seem or appropriate unless you are testing the error handling.

File: regression/abstraction/test2/Readme [hunk 8]
```diff
@@ -0,0 +1,4 @@
+Assertions with function "bar" that has hidden index variable.
+We error out currently because functions are not supported inside assertions. 
+But even when they are, we should still error as index variable doesn't appear explicitly in the
+args list and it cannot be checked using is_precise function.
```
Comment (martin-cs) [inline regression/abstraction/test2/Readme]: Rather than a readme, why not put this in the comments section of the `test.desc`?

File: regression/abstraction/test3/Readme [hunk 9]
```diff
@@ -0,0 +1,2 @@
+Assertion has function "bar" which has index it uses explicitly in the signature.
```
Comment (martin-cs) [inline regression/abstraction/test3/Readme]: Again; in the `test.desc` please.

File: src/cbmc/cbmc_parse_options.cpp [hunk 10]
```diff
@@ -75,6 +76,8 @@ Author: Daniel Kroening, kroening@kroening.com
 #include <goto-instrument/nondet_static.h>
 #include <goto-instrument/reachability_slicer.h>
 
+#include <goto-instrument/abstraction_spect.h>
```
Comment (martin-cs) [inline src/cbmc/cbmc_parse_options.cpp]: Check the coding standard but there was an attempt to have the headers sorted alphabetically.  Certainly with the other `goto-instrument` includes would be good.

File: src/cbmc/cbmc_parse_options.cpp [hunk 11]
```diff
@@ -804,6 +812,21 @@ int cbmc_parse_optionst::get_goto_program(
...
+  {
+    std::string abst_file = options.get_option("use-rra");
+
+    abstraction_spect abst_info(abst_file, ui_message_handler);
+
+    std::vector<std::string> abstfiles =
+      abst_info.get_abstraction_function_files();
+
+    goto_modelt goto_model_for_abst_fns =
+      initialize_goto_model(abstfiles, ui_message_handler, options);
```
Comment (martin-cs) [inline src/cbmc/cbmc_parse_options.cpp]: `use-rra` needs a re-initialize and re-link?

File: src/cbmc/cbmc_parse_options.cpp [hunk 12]
```diff
@@ -1137,6 +1160,10 @@ void cbmc_parse_optionst::help()
     HELP_TIMESTAMP
     " --write-solver-stats-to json-file\n"
     "                              collect the solver query complexity\n"
+    // NOLINTNEXTLINE(whitespace/line_length)
+    " --use-rra abstraction-file\n"
+    // NOLINTNEXTLINE(whitespace/line_length)
+    "                              abstract the program as specified in the abstraction-file"
```
Comment (martin-cs) [inline src/cbmc/cbmc_parse_options.cpp]: Too generic, how about `apply the replication reducing abstraction configured by the abstraction-file`?

File: src/cbmc/cbmc_parse_options.h [hunk 13]
```diff
@@ -83,6 +83,7 @@ class optionst;
   "(round-to-nearest)(round-to-plus-inf)(round-to-minus-inf)(round-to-zero)" \
   OPT_FLUSH \
   "(localize-faults)" \
+  "(use-abstraction):" \
```
Comment (martin-cs) [inline src/cbmc/cbmc_parse_options.h:86]: ??? Do the tests actually work?

File: src/goto-instrument/Makefile [hunk 14]
```diff
@@ -1,4 +1,5 @@
-SRC = accelerate/accelerate.cpp \
+SRC = abstraction.cpp \
```
Comment (martin-cs) [inline src/goto-instrument/Makefile:1]: Again, too generic, `replication_reducing_abstraction.cpp` ?

File: src/goto-instrument/Makefile [hunk 15]
```diff
@@ -12,6 +13,7 @@ SRC = accelerate/accelerate.cpp \
       accelerate/scratch_program.cpp \
       accelerate/trace_automaton.cpp \
       accelerate/util.cpp \
+      abstraction_spect.cpp \
```
Comment (martin-cs) [inline src/goto-instrument/Makefile:16]: Similar.

File: src/goto-instrument/abstraction.cpp [hunk 16]
```diff
@@ -0,0 +1,45 @@
+/*******************************************************************\
+
+Module: Abstraction
```
Comment (martin-cs) [inline src/goto-instrument/abstraction.cpp]: Too ambiguous.

File: src/goto-instrument/abstraction.h [hunk 17]
```diff
@@ -0,0 +1,43 @@
+/*******************************************************************\
+
+Module: Abstraction
```
Comment (martin-cs) [inline src/goto-instrument/abstraction.h]: Again, too ambiguous.

File: src/goto-instrument/abstraction.h [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+Author: Lefan Zhang, lefanz@amazon.com
+        Murali Talupur talupur@amazon.com
+
+\*******************************************************************/
+
+/// \file
+/// Replication Reducing Abstraction (RRA) for getting unbounded proofs.
+/// User can specify a json file with arrays/strings to be abstracted
+/// along with shapes to be used. See regression/abstraction folder
+/// for some examples.
```
Comment (martin-cs) [inline src/goto-instrument/abstraction.h]: More here please.

File: src/goto-instrument/abstraction_spect.h [hunk 19]
```diff
@@ -0,0 +1,482 @@
...
+#define CPROVER_GOTO_INSTRUMENT_ABSTRACTION_SPECT_H
+
+#include <limits>
+#include <list>
+#include <map>
+#include <string>
+#include <tuple>
+#include <vector>
+
+class abstraction_spect
```
Comment (martin-cs) [inline src/goto-instrument/abstraction_spect.h]: Too ambigious. `rr_abstraction_spect`?

File: src/goto-instrument/goto_instrument_parse_options.cpp [hunk 20]
```diff
@@ -24,12 +24,16 @@ Author: Daniel Kroening, kroening@kroening.com
 #include <util/unicode.h>
 #include <util/version.h>
 
+#include <json/json_parser.h>
```
Comment (martin-cs) [inline src/goto-instrument/goto_instrument_parse_options.cpp]: Is this needed?

File: src/goto-instrument/goto_instrument_parse_options.cpp [hunk 21]
```diff
@@ -1012,6 +1018,21 @@ void goto_instrument_parse_optionst::instrument_goto_program()
   else
     options.set_option("assert-to-assume", false);
 
+  if(cmdline.isset("use-abstraction"))
```
Comment (martin-cs) [inline src/goto-instrument/goto_instrument_parse_options.cpp]: Command line should be consistent with CBMC.

File: src/goto-instrument/goto_instrument_parse_options.cpp [hunk 22]
```diff
@@ -1012,6 +1018,21 @@ void goto_instrument_parse_optionst::instrument_goto_program()
...
+  {
+    std::string abs_file = cmdline.get_value("use-abstraction");
```
Comment (martin-cs) [inline src/goto-instrument/goto_instrument_parse_options.cpp]: If this code is common to both CBMC and goto-instrument, why not put it all in a function so that there is not a parallel maintainence issue?

File: src/goto-instrument/goto_instrument_parse_options.cpp [hunk 23]
```diff
@@ -1779,6 +1800,8 @@ void goto_instrument_parse_optionst::help()
     " --drop-unused-functions      drop functions trivially unreachable from main function\n" // NOLINT(*)
     " --print-internal-representation\n" // NOLINTNEXTLINE(*)
     "                              show verbose internal representation of the program\n"
+    " --print-detailed-ir\n" // NOLINTNEXTLINE(*)
```
Comment (martin-cs) [inline src/goto-instrument/goto_instrument_parse_options.cpp]: Are there tests for this?

File: src/goto-instrument/goto_instrument_parse_options.cpp [hunk 24]
```diff
@@ -1793,6 +1816,9 @@ void goto_instrument_parse_optionst::help()
     " --show-local-safe-pointers   show pointer expressions that are trivially dominated by a not-null check\n" // NOLINT(*)
     " --show-safe-dereferences     show pointer expressions that are trivially dominated by a not-null check\n" // NOLINT(*)
     "                              *and* used as a dereference operand\n" // NOLINT(*)
+    " --show-index-exprs array_name\n"
```
Comment (martin-cs) [inline src/goto-instrument/goto_instrument_parse_options.cpp]: Again, this appears unrelated.

File: src/cbmc/cbmc_parse_options.h (hunk 13 above)
Comment (martin-cs) [reply src/cbmc/cbmc_parse_options.h:86]: As the option is shared with goto-instrument, could we have `OPT_` and `HELP_` macros?

File: src/goto-instrument/module_dependencies.txt [hunk 25]
```diff
@@ -5,6 +5,7 @@ assembler
 cpp
 goto-instrument
 goto-programs
+json
```
Comment (martin-cs) [inline src/goto-instrument/module_dependencies.txt:8]: Needed?

File: src/util/expr.h [hunk 26]
```diff
@@ -84,6 +84,31 @@ class exprt:public irept
     return static_cast<const typet &>(find(ID_type));
   }
 
+  static void indent_str(std::string &s, unsigned indent)
+  {
+    for(unsigned i = 0; i < indent; i++)
+      s += ' ';
+  }
+
+  const std::string get_type_string(int depth) const
```
Comment (martin-cs) [inline src/util/expr.h]: How does this differ to just `e.type().pretty()` with the appropriate indent arguments?

File: src/util/file_util.cpp [hunk 27]
```diff
@@ -77,6 +78,43 @@ std::string get_current_working_directory()
   return working_directory;
 }
 
+/// \param rel_path: relative path
+/// \return absolute path
+std::string get_absolute_path(const std::string &rel_path)
+{
+#ifndef _WIN32
+  errno = 0;
+  char rp[4096];
```
Comment (martin-cs) [inline src/util/file_util.cpp]: Why 4096?  If it is platform specific then use the macro for it.

File: unit/Makefile [hunk 28]
```diff
@@ -168,6 +168,7 @@ BMC_DEPS =../src/cbmc/c_test_input_generator$(OBJEXT) \
           ../src/goto-cc/armcc_cmdline$(OBJEXT) \
           ../src/goto-cc/goto_cc_cmdline$(OBJEXT) \
           ../src/goto-instrument/source_lines$(OBJEXT) \
+          ../src/goto-instrument/abstraction_spect$(OBJEXT) \
```
Comment (martin-cs) [inline unit/Makefile]: ??? I think the unit tests are missing?  Also `rr_abstraction_spect` or `replication_reducting_abstraction_spect`?

**Review by martin-cs — CHANGES_REQUESTED** [review]
I am intrigued to know more about this but there are changes needed.

File: regression/abstraction/Makefile (hunk 1 above)
Comment (muralitaws) [reply regression/abstraction/Makefile]: Ok, I will change it to rr-abstraction.

File: regression/abstraction/abstraction.sh (hunk 5 above)
Comment (muralitaws) [reply regression/abstraction/abstraction.sh]: Ok, will rename abstraction --> rr-abstraction.

File: regression/abstraction/abstraction.sh (hunk 6 above)
Comment (muralitaws) [reply regression/abstraction/abstraction.sh]: This number comes from the particulars of our approach. For examples considered here, this unwinding number is sufficient with rr-abstraction.

File: regression/abstraction/test2/Readme (hunk 7 above)
Comment (muralitaws) [reply regression/abstraction/test2/Readme]: For the functionality in this PR we don't error out. The readme is referring to the state as it is in our local repo. I will delete the readme files from this PR and introduce them in the next one.

File: regression/abstraction/test2/Readme (hunk 8 above)
Comment (muralitaws) [reply regression/abstraction/test2/Readme]: Sure, I can do that.

File: regression/abstraction/test3/Readme (hunk 9 above)
Comment (muralitaws) [reply regression/abstraction/test3/Readme]: Ok.

File: src/cbmc/cbmc_parse_options.cpp (hunk 10 above)
Comment (muralitaws) [reply src/cbmc/cbmc_parse_options.cpp]: We have decided to have --use-rra option only for goto-instrument, not for cbmc. This code disappears in the next commit.

File: src/cbmc/cbmc_parse_options.cpp (hunk 11 above)
Comment (muralitaws) [reply src/cbmc/cbmc_parse_options.cpp]: This code is removed now.

File: src/cbmc/cbmc_parse_options.cpp (hunk 12 above)
Comment (muralitaws) [reply src/cbmc/cbmc_parse_options.cpp]: This option is removed.

File: src/cbmc/cbmc_parse_options.h (hunk 13 above)
Comment (muralitaws) [reply src/cbmc/cbmc_parse_options.h:86]: Yes, the regression tests do work because they are using the option in goto-instrument. Cbmc option is not used, so we have dropped it completely.

File: src/goto-instrument/Makefile (hunk 14 above)
Comment (muralitaws) [reply src/goto-instrument/Makefile:1]: Renamed to rr-abstraction.cpp.

File: src/goto-instrument/Makefile (hunk 15 above)
Comment (muralitaws) [reply src/goto-instrument/Makefile:16]: Renamed to rr-abstraction-spect.cpp

File: src/goto-instrument/abstraction.cpp (hunk 16 above)
Comment (muralitaws) [reply src/goto-instrument/abstraction.cpp]: Renamed.

File: src/goto-instrument/abstraction.h (hunk 17 above)
Comment (muralitaws) [reply src/goto-instrument/abstraction.h]: Renamed.

File: src/goto-instrument/abstraction.h (hunk 18 above)
Comment (muralitaws) [reply src/goto-instrument/abstraction.h]: Ok.

File: src/goto-instrument/abstraction_spect.h (hunk 19 above)
Comment (muralitaws) [reply src/goto-instrument/abstraction_spect.h]: Done.

File: src/goto-instrument/goto_instrument_parse_options.cpp (hunk 20 above)
Comment (muralitaws) [reply src/goto-instrument/goto_instrument_parse_options.cpp]: Removed.

File: src/goto-instrument/goto_instrument_parse_options.cpp (hunk 21 above)
Comment (muralitaws) [reply src/goto-instrument/goto_instrument_parse_options.cpp]: Changed.

File: src/goto-instrument/goto_instrument_parse_options.cpp (hunk 22 above)
Comment (muralitaws) [reply src/goto-instrument/goto_instrument_parse_options.cpp]: Removed the option from CBMC completely.

File: src/goto-instrument/goto_instrument_parse_options.cpp (hunk 23 above)
Comment (muralitaws) [reply src/goto-instrument/goto_instrument_parse_options.cpp]: We were using this for our internal tests. It shouldn't be in the PR, removed it.

File: src/goto-instrument/goto_instrument_parse_options.cpp (hunk 24 above)
Comment (muralitaws) [reply src/goto-instrument/goto_instrument_parse_options.cpp]: This too was for our internal tests. Removed it from the PR.

File: src/goto-instrument/module_dependencies.txt (hunk 25 above)
Comment (muralitaws) [reply src/goto-instrument/module_dependencies.txt:8]: Yes, this is needed.

File: src/util/expr.h (hunk 26 above)
Comment (muralitaws) [reply src/util/expr.h]: This code is now removed. Was part of the print-detailed-ir option.

File: unit/Makefile (hunk 28 above)
Comment (muralitaws) [reply unit/Makefile]: This dependency is no longer needed. Changed  to rr_abstraction_spect.

File: regression/abstraction/Makefile (hunk 1 above)
Comment (muralitaws) [reply regression/abstraction/Makefile]: Changed to rr-abstraction.

File: regression/abstraction/abst-funcs.c (hunk 3 above)
Comment (muralitaws) [reply regression/abstraction/abst-funcs.c]: Changed.

File: regression/abstraction/abst-funcs.c (hunk 4 above)
Comment (muralitaws) [reply regression/abstraction/abst-funcs.c]: That works too. Changed.

File: regression/abstraction/abstraction.sh (hunk 5 above)
Comment (muralitaws) [reply regression/abstraction/abstraction.sh]: Changed it to rr-abstraction.sh.

File: regression/abstraction/abst-funcs.c (hunk 2 above)
Comment (muralitaws) [reply regression/abstraction/abst-funcs.c]: Not sure I understand this. There is no header for abst-funcs.c. It's there for rr-abstraction to choose abstraction functions from.

File: src/goto-instrument/goto_instrument_parse_options.cpp (hunk 24 above)
Comment (muralitaws) [reply src/goto-instrument/goto_instrument_parse_options.cpp]: Removed this option from PR.

File: src/goto-instrument/module_dependencies.txt (hunk 25 above)
Comment (muralitaws) [reply src/goto-instrument/module_dependencies.txt:8]: Yes, we get cpplint errors without this.

File: src/util/expr.h (hunk 26 above)
Comment (muralitaws) [reply src/util/expr.h]: This code is removed in the next commit.

File: src/util/file_util.cpp (hunk 27 above)
Comment (muralitaws) [reply src/util/file_util.cpp]: Changed the code to remove this.

File: regression/abstraction/abst-funcs.c (hunk 2 above)
Comment (martin-cs) [reply regression/abstraction/abst-funcs.c]: I think the problem might be that there is no explanation of how this works.  My point is that if you are manually modelling things, simply having:

```
size_t nondet_size_t(void);
```

is sufficient and gives you the same behaviour.

File: regression/abstraction/abstraction.sh (hunk 6 above)
Comment (martin-cs) [reply regression/abstraction/abstraction.sh]: If it is specific to this example then that should be documented and explained.  If it is required for all uses of `--rr-abstraction` then maybe it could be set in the command-line argument parsing?

File: regression/abstraction/test2/Readme (hunk 7 above)
Comment (martin-cs) [reply regression/abstraction/test2/Readme]: Yeah; at the end of each PR things need to be in a consistent state, including documentation and test status.

File: regression/abstraction/test2/Readme (hunk 8 above)
Comment (martin-cs) [reply regression/abstraction/test2/Readme]: Thanks!

File: src/cbmc/cbmc_parse_options.cpp (hunk 10 above)
Comment (martin-cs) [reply src/cbmc/cbmc_parse_options.cpp]: Then please remove it from the PR.  Ideally each PR we merge should be a finished thing and not depend on others.

File: src/goto-instrument/Makefile (hunk 15 above)
Comment (martin-cs) [reply src/goto-instrument/Makefile:16]: Great; thank you.  I appreciate that this is kinda tedious but there are at least ... 4, maybe more, different kinds of abstraction in the code-base already.

File: src/goto-instrument/module_dependencies.txt (hunk 25 above)
Comment (martin-cs) [reply src/goto-instrument/module_dependencies.txt:8]: Interesting...  Thanks.

File: src/util/expr.h (hunk 26 above)
Comment (martin-cs) [reply src/util/expr.h]: Can you squash the removals so that this PR doesn't even touch that file?  It reduces the number of reviews / different people you need to get sign-off from.

File: regression/rr-abstraction/test8/readme [hunk 29]
```diff
@@ -0,0 +1,9 @@
+Copied from test1.
+Main.c has a simple while loop that copies a2 array into a1 but it also has "sum" 
+which is not an abstracted variable that is being assigned inside the array. This should 
+lead to an error.
+
+With spec2.json there should be no error.
+Value of "sum" is not dependent on the loop size.
+
+Spec1.json is not used.
```
Comment (martin-cs) [inline regression/rr-abstraction/test8/readme:9]: I think these `readme` files should be in the `test.desc`.

File: src/cbmc/Makefile [hunk 30]
```diff
@@ -23,6 +23,7 @@ OBJ += ../ansi-c/ansi-c$(LIBEXT) \
       ../pointer-analysis/rewrite_index$(OBJEXT) \
       ../pointer-analysis/goto_program_dereference$(OBJEXT) \
       ../goto-instrument/source_lines$(OBJEXT) \
+      ../goto-instrument/abstraction_spect$(OBJEXT) \
```
Comment (martin-cs) [inline src/cbmc/Makefile:26]: If you have removed this from CBMC, why is this needed?

File: src/cbmc/cbmc_parse_options.cpp [hunk 31]
```diff
@@ -51,6 +51,7 @@ Author: Daniel Kroening, kroening@kroening.com
 #include <goto-programs/adjust_float_expressions.h>
 #include <goto-programs/initialize_goto_model.h>
 #include <goto-programs/instrument_preconditions.h>
+#include <goto-programs/link_goto_model.h>
```
Comment (martin-cs) [inline src/cbmc/cbmc_parse_options.cpp:54]: Again... why is this needed?

File: src/cbmc/cbmc_parse_options.h (hunk 13 above)
Comment (martin-cs) [reply src/cbmc/cbmc_parse_options.h:86]: It still seems to appear in the PR...

**Review by martin-cs — CHANGES_REQUESTED** [review]
Still lacking clear documentation and the regression tests adding to `regression/Makefile`.

**Comment by muralitaws** [comment]
I am making this PR a draft while I work on adding more explanation for how the abstraction works. That will answer a lot of questions and also make reviewing future PRs smoother.

**Comment by martin-cs** [comment]
@muralitalupur : thank sounds good.  Thanks for pushing this.

**Comment by feliperodri** [comment]
@zlfben will conduct this work moving forward.
